crates/sui-framework/packages/sui-system/sources/validator_set.move

```diff
@@ -233,11 +233,37 @@ module sui_system::validator_set {
             EDuplicateValidator
         );
         assert!(validator.is_preactive(), EValidatorNotCandidate);
-        assert!(validator.total_stake_amount() >= min_joining_stake_amount, EMinJoiningStakeNotReached);
+        assert!(can_join(self, validator.total_stake_amount(), ctx), EMinJoiningStakeNotReached);
 
         self.pending_active_validators.push_back(validator);
     }
 
+    /// Return `true` if a  candidate validator with `stake` will have sufficeint voting power to join the validator set
+    fun can_join(self: &ValidatorSet, stake: u64, ctx: &TxContext): bool {
+        let (min_joining_voting_power, _, _) = get_voting_power_thresholds(ctx);
+        // if the validator will have at least `min_joining_voting_power` after joining, they can join.
+        // this formula comes from SIP-39 TODO link once landed
+        let future_total_stake = self.total_stake + stake;
+        let future_validator_voting_power = voting_power::derive_voting_power(stake, future_total_stake);
+        future_validator_voting_power >= min_joining_voting_power
+    }
+
+    /// return (min, low, very low voting power) thresholds
+    fun get_voting_power_thresholds(ctx: &TxContext): (u64, u64, u64) {
```

Comment:
@lxfind, @emmazzz told me it would be ok/preferable to use module constants for these instead of `SuiSystemStateV2`. So the way I handled this is marking the old stake-based constants in `SuiSystemV2` as deprecated, and using this as the source of truth for the new voting-power based constants. Let me know if that doesn't make sense, and if there's anything should change in the Rust version of `SuiSystemV2`.

Reply:
Do we need different values for these in different networks?